api/include/opentelemetry/trace/link.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <iostream>
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/trace/key_value_iterable_view.h"
+#include "opentelemetry/trace/span_context.h"
+#include "opentelemetry/version.h"
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace trace
+{
+class Link final
+{
+public:
+  Link() = default;
+
+  Link(const opentelemetry::trace::SpanContext &span_context,
+       const trace_api::KeyValueIterable &attributes)
+      : span_context_(span_context), attributes_(attributes)
+  {}
+
+  template <class T, nostd::enable_if_t<detail::is_key_value_iterable<T>::value> * = nullptr>
+  Link(const SpanContext &span_context, const T &attributes)
+      : Link(span_context, KeyValueIterableView<T>(attributes))
+  {}
+
+  Link(const SpanContext &span_context,
+       std::initializer_list<std::pair<nostd::string_view, common::AttributeValue>> attributes)
+      : Link(span_context,
+             nostd::span<const std::pair<nostd::string_view, common::AttributeValue>>{
+                 attributes.begin(), attributes.end()})
+  {}
+
+  Link(const opentelemetry::trace::SpanContext &span_context)
+      : Link(span_context,
+             nostd::span<const std::pair<nostd::string_view, common::AttributeValue>>{})
+  {}
+
+  // returns the Span Context associated with this Link
+  const SpanContext &GetSpanContext() const { return span_context_; }
+
+  // returns the attributes associated with link
+  const KeyValueIterable &GetAttributes() const { return attributes_; }
+
+private:
+  const trace_api::SpanContext &span_context_;
+  const trace_api::KeyValueIterable &attributes_;
```

Comment:
A `KeyValueIterable` is similar to a non-owning reference (like `nostd::span`). Having it as a class member will increase the possibility for many subtle bugs. This test, for example, crashes:
```c++
TEST(LinkTest, CreateLinkIterableViewInvalid)
{
  auto m1 = new std::map<std::string, std::string>();
  m1->insert(std::pair<std::string, std::string>("abc", "123"));
  m1->insert(std::pair<std::string, std::string>("xyz", "456"));
  SpanContext s1(true, true);

  Link link(s1, *m1);
  delete m1;

  std::map<nostd::string_view, nostd::string_view> m2;
  m2["abc"] = "123";
  m2["xzy"] = "456";
  validateKeyValueIterableView(link.GetAttributes(), m2);
  EXPECT_EQ(s1.IsSampled(), link.GetSpanContext().IsSampled());
}
```

Reply:
Thanks for pointing it out. Let me look into that.

Reply:
I actually think we should never force the user to explicitly create a `Link` object, but rather provide _views_ that  are automatically constructed from arguments, like we are doing with `KeyValueIterableView`. I came up with a prototype API for that, you can see it in [this diff](https://github.com/open-telemetry/opentelemetry-cpp/compare/master...pyohannes:span-links?expand=1).

This would allow us to accept link arguments from standard containers:
```c++
std::map<std::string, std::string> attrs                                      = {{"a", "3"}};
std::vector<std::pair<SpanContext, std::map<std::string, std::string>>> links = {
      {SpanContext(), attrs}};
auto s1 = tracer->StartSpan("abc", attrs, links);
```

Or from initializer lists:
```c++
auto s2 = tracer->StartSpan("efg", {{"a", 3}}, {{SpanContext(), {{"b", 4}}}});
```
Either way, we accept initializer lists or containers, which relieves us from many risks that we run into when we force the user to explicitly construct a `Link` object (like the crash mentioned above).

Please let me know your thoughts, or whether you have any questions about that approach.

Reply:
Agree -  the iterable approach as we do with `KeyValueIterableView` looks much cleaner solution. Please go ahead and raise the PR, I can then modify the `sdk` to add links to the span as part of this PR. Or if you want me to incorporate these changes as part of this PR, do let me know. I would be fine either way.

Reply:
I will not have time to polish my solution and submit a PR before next week. So feel free to incorporate those changes into this PR. Otherwise I'll do it next week.

Reply:
No problem, I should be able to incorporate these in current PR during this week. Thanks.

Reply:
@pyohannes  - I have incorporated the changes as you suggested [here](https://github.com/open-telemetry/opentelemetry-cpp/compare/master...lalitb:span-link-3?expand=1). I still would like to improve `ForEachKeyValue` function [here](https://github.com/lalitb/opentelemetry-cpp/blob/9d078ac4d1c18fd2609fbe2f5dc1690e2e04cb43/api/include/opentelemetry/trace/span_context_kv_iterable_view.h#L55) to iterate over `SpanContext` and `KeyValueIterble`, as that would help me avoid temporarily creating unordered_map in [`span.cc`](https://github.com/lalitb/opentelemetry-cpp/blob/9d078ac4d1c18fd2609fbe2f5dc1690e2e04cb43/sdk/src/trace/span.cc#L106) to further pass to `recordable`. Please let me know your feedback on the changes, and whether we can achieve the improvement I am thinking.

Reply:
@pyohannes  - I have now incorporated above changes in current PR, that would help reviewing above mentioned issues better. thanks.

Reply:
> I still would like to improve ForEachKeyValue function here to iterate over SpanContext and KeyValueIterble, as that would help me avoid temporarily creating unordered_map in span.cc to further pass to recordable. 

Yes, I fully agree. Adapting the callback to accept a `KeyValueIterable` would make the code a lot simpler.

Reply:
@pyohannes  - This is fixed now, can you please re-review it now. thanks.